src/NHibernate/Dialect/SQLiteDialect.cs

```diff
@@ -490,8 +490,10 @@ protected class SQLiteCastFunction : CastFunction
 		{
 			protected override bool CastingIsRequired(string sqlType)
 			{
-				// SQLite doesn't support casting to datetime types.  It assumes you want an integer and destroys the date string.
-				if (StringHelper.ContainsCaseInsensitive(sqlType, "date") || StringHelper.ContainsCaseInsensitive(sqlType, "time"))
+				// SQLite doesn't support casting to datetime or uniqueidentifier types. It assumes you want an integer and destroys the date or uniqueidentifier string.
```

Comment:
```suggestion
				// SQLite does not have types for datetime, date, time, and uniqueidentifier, they are to be handled as text. Trying to cast to them causes the SQLite lax type inference from unknown type to interpret the cast as casting to numeric, destroying the value. See https://www.sqlite.org/datatype3.html#determination_of_column_affinity.
```

Reply:
> Trying to cast to them causes the SQLite lax type inference from unknown type to interpret the cast as casting to numeric, destroying the value

That's some tough sentence )  Original version seems easier to comprehend...

Reply:
The trouble for me is, the previous sentence is somewhat wrong. It somewhat implies SQLite has a datetime or uniqueidentifier type. But no, it has not.

It seems there are some misunderstanding in NHibernate about SQLite type system, with invalid types used in dialect, like:
https://github.com/nhibernate/nhibernate-core/blob/708ae88ab544523ab9fa8d277483a7131619d7b7/src/NHibernate/Dialect/SQLiteDialect.cs#L39-L67
This method is quite rubbish, since the only actual types handled by SQLite are `text`, `real`, `integer`, `blob` and `numeric`. Using any other type name does not cause any exception but causes SQLite to infer the type with some brittle logic:

> 1. If the declared type contains the string "INT" then it is assigned INTEGER affinity.
> 2. If the declared type of the column contains any of the strings "CHAR", "CLOB", or "TEXT" then that column has TEXT affinity. Notice that the type VARCHAR contains the string "CHAR" and is thus assigned TEXT affinity.
> 3. If the declared type for a column contains the string "BLOB" or if no type is specified then the column has affinity BLOB.
> 4. If the declared type for a column contains any of the strings "REAL", "FLOA", or "DOUB" then the column has REAL affinity.
> 5. Otherwise, the affinity is NUMERIC.

This means the last mappings, for date, time, bool, uniqueidentifier, all ends-up being treated as `numeric`. Which generally does not fail but is inefficient: SQLite will attempt to convert each value in such case, and falls back to storing them as text if not convertible.

Back to the comment: in fact I should even have written it this way:

> NHibernate dialect type bindings for SQLite are invalid and cause convert to have unexpected results when attempting to convert date, time and uniqueidentifier.